functions/index.js

```diff
@@ -0,0 +1,277 @@
+
+      if (userId in userIdObject) {
+        const pushToken = userIdObject[userId]['token'];
+        console.log(`pushToken FOUND for userId ${userId}`);
+        console.log(pushToken);
+
+        // See the "Defining the message payload" section below for details
+        // on how to define a message payload.
+        // NOTE: the notification dictionary HAS TO BE HERE
+        // read this: https://firebase.google.com/docs/cloud-messaging/concept-options#notifications
+
+        // something like this:
+        //   var payload = {
+          //   notification: {
+          //     title: 'WappuNotif!',
+          //     body: 'jep :-)'
+          //   },
+          //   data: {
+          //     kv1: 'any data here',
+          //     kv2: 'any data here, too'
+          //   }
+          // };
+
+        // Send a message to the device corresponding to the provided
+        // registration token.
+        admin.messaging().sendToDevice(pushToken, payload)
+          .then(function(response) {
+            // See the MessagingDevicesResponse reference documentation for
+            // the contents of response.
+            console.log("Successfully sent message:", response);
+          })
+          .catch(function(error) {
+            console.log("Error sending message:", error);
+          });
+
+      } else {
+        console.log(`userId ${userId} did not have pushToken... abort sending!`);
+      }
+    })
+};
+
+// THIS TRIGGERS ON WRITE @ /chats/whatever
+// figure out who should receive a message
+// call the message sending logic
+// the receiver of this message is the other person in the chat
+// ----> first get the new message's sender
+// ----> figure out the second party in the chat
+exports.sendPushMessage = functions.database.ref('/chats/{chatId}/')
+  .onWrite(event => {
+    // Grab the current value of what was written to the Realtime Database.
+    const chatData = event.data.val();
+    console.log('\ŋ\ŋthere was a new write to ' + event.params.chatId);
+    console.log(chatData)
+    const users = chatData.users;
+    const lastMessageId = Object.keys(chatData.messages).slice(-1)[0];
+    const userIdForSender = chatData.messages[lastMessageId].userId;
+    const userIdForReceiver = users.filter(id => id !== userIdForSender)[0];
+    console.log(users)
+    console.log(userIdForSender)
+    console.log(userIdForReceiver)
+
+    const payload = {
+      notification: {
+        title: 'Whappu!',
+        body: 'You have a new message!'
+      }
+    };
+    return updateReadReceiptAndSendPushNotification(userIdForReceiver, 'msg', payload);
+})
+
+// this writes "closed": true to chats/chatId and so closes the chat
+// for writing. reading will still be allowed.
+exports.closeChatId = functions.https.onRequest((req, res) => {
+
+  if (req.get('FUNCTION_SECRET_KEY') !== functions.config().functions.secret) {
+    console.log('not authenticated, FUNCTION_SECRET_KEY header is barps');
+    res.sendStatus(403);
```

Comment:
A 401 status code would most likely be more fitting here, since we are refusing authentication. 403 status codes are used when there's an authorization issue (ie. you are logged in / using a valid token, but you don't have access to this resource)